C error: cannot convert 'struct array' to 'struct string *'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.7 7baff15

CC version: cc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
thirdparty/tcc: N/A

```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg liveview_bug.v`
```v


type Either = string | int

fn tag(params ...Either) int {
	return 0
}

fn div(params ...Either) int {
	return tag(params)
}


```

**What did you expect to see?**

compilation complete ok

**What did you see instead?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/liveview_bug.01J8NVCRNQZEBSJ5FMBMQZ2DW7.tmp.c:13255: error: cannot convert 'struct array' to 'struct string *'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2607,8 +2607,9 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 						}
 					}
 				} else {
-					if args.len == 1
-						&& g.table.type_to_str(arr_info.elem_type) == g.table.type_to_str(varg_type) {
+					// passing variadic arg to another call which expects same array type
+					if args.len == 1 && args[arg_nr].typ.has_flag(.variadic)
+						&& args[arg_nr].typ == varg_type {
 						g.ref_or_deref_arg(args[arg_nr], arr_info.elem_type, node.language,
 							false)
 					} else {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2607,16 +2607,22 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 						}
 					}
 				} else {
-					noscan := g.check_noscan(arr_info.elem_type)
-					g.write('new_array_from_c_array${noscan}(${variadic_count}, ${variadic_count}, sizeof(${elem_type}), _MOV((${elem_type}[${variadic_count}]){')
-					for j in arg_nr .. args.len {
-						g.ref_or_deref_arg(args[j], arr_info.elem_type, node.language,
+					if args.len == 1
+						&& g.table.type_to_str(arr_info.elem_type) == g.table.type_to_str(varg_type) {
+						g.ref_or_deref_arg(args[arg_nr], arr_info.elem_type, node.language,
 							false)
-						if j < args.len - 1 {
-							g.write(', ')
+					} else {
+						noscan := g.check_noscan(arr_info.elem_type)
+						g.write('new_array_from_c_array${noscan}(${variadic_count}, ${variadic_count}, sizeof(${elem_type}), _MOV((${elem_type}[${variadic_count}]){')
+						for j in arg_nr .. args.len {
+							g.ref_or_deref_arg(args[j], arr_info.elem_type, node.language,
+								false)
+							if j < args.len - 1 {
+								g.write(', ')
+							}
 						}
+						g.write('}))')
 					}
-					g.write('}))')
 				}
 			} else {
 				g.write('__new_array(0, 0, sizeof(${elem_type}))')

--- a/vlib/v/tests/fns/variadic_sumtype_test.v
+++ b/vlib/v/tests/fns/variadic_sumtype_test.v
@@ -1,0 +1,19 @@
+type Either = string | int
+
+fn tag(params ...Either) int {
+	mut c := 0
+	for p in params {
+		if p is int {
+			c += p
+		}
+	}
+	return c
+}
+
+fn div(params ...Either) int {
+	return tag(params)
+}
+
+fn test_main() {
+	assert dump(div('foo', 1, 2)) == 3
+}

